azurerm/internal/services/datafactory/resource_arm_data_factory_integration_runtime_self_hosted.go

```diff
@@ -0,0 +1,269 @@
+				},
+				Set: resourceArmDataFactoryIntegrationRuntimeSelfHostedRbacAuthorizationHash,
+			},
+		},
+	}
+}
+
+func resourceArmDataFactoryIntegrationRuntimeSelfHostedCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).DataFactory.IntegrationRuntimesClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	factoryName := d.Get("data_factory_name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	if features.ShouldResourcesBeImported() && d.IsNewResource() {
+		existing, err := client.Get(ctx, resourceGroup, factoryName, name, "")
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("Error checking for presence of existing Data Factory Self-Hosted Integration Runtime %q (Resource Group %q, Data Factory %q): %s", name, resourceGroup, factoryName, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_data_factory_integration_runtime_self_hosted", *existing.ID)
+		}
+	}
+
+	description := d.Get("description").(string)
+
+	selfHostedIntegrationRuntime := datafactory.SelfHostedIntegrationRuntime{
+		Description: &description,
+		Type:        datafactory.TypeSelfHosted,
+		SelfHostedIntegrationRuntimeTypeProperties: expandAzureRmDataFactoryIntegrationRuntimeSelfHostedTypeProperties(d),
+	}
+
+	basicIntegrationRuntime, _ := selfHostedIntegrationRuntime.AsBasicIntegrationRuntime()
+
+	integrationRuntime := datafactory.IntegrationRuntimeResource{
+		Name:       &name,
+		Properties: basicIntegrationRuntime,
+	}
+
+	if _, err := client.CreateOrUpdate(ctx, resourceGroup, factoryName, name, integrationRuntime, ""); err != nil {
+		return fmt.Errorf("Error creating/updating Data Factory Self-Hosted Integration Runtime %q (Resource Group %q, Data Factory %q): %+v", name, resourceGroup, factoryName, err)
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, factoryName, name, "")
+	if err != nil {
+		return fmt.Errorf("Error retrieving Data Factory Self-Hosted Integration Runtime %q (Resource Group %q, Data Factory %q): %+v", name, resourceGroup, factoryName, err)
+	}
+
+	if resp.ID == nil {
+		return fmt.Errorf("Cannot read Data Factory Self-Hosted Integration Runtime %q (Resource Group %q, Data Factory %q) ID", name, resourceGroup, factoryName)
+	}
+
+	d.SetId(*resp.ID)
+
+	return resourceArmDataFactoryIntegrationRuntimeSelfHostedRead(d, meta)
+}
+
+func resourceArmDataFactoryIntegrationRuntimeSelfHostedRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).DataFactory.IntegrationRuntimesClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	factoryName := id.Path["factories"]
+	name := id.Path["integrationruntimes"]
```

Comment:
and a resource specific parse ID function here

Reply:
None of these are IDs they are names.  So there isn't really any path to parse.  Both name and data_factory_name are validated via the schema.

Reply:
Hi @pearcec 
I think Kt was referring to the pattern of a custom parser for the whole resource ID (e.g. `/subscriptions/12345678-1234-1234-1234-12345678abc/resourceGroups/exampleResourceGroup/providers/Microsoft.DataFactory/factories/exampleFactoryName/integrationruntimes/exampleIntegrationRuntime` which could be parsed into a custom type, and then accessed as `id.FactoryName` and `id.Name` - (example [here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/compute/parse/image.go))

Reply:
ok makes sense, I will take a look.

Reply:
Ok I added it.